### ISSUE TITLE
clpb missing a solution

### ISSUE THREAD

**Issue by infradig** [trealla-prolog/trealla#338]
The only example from the clpb.pl doc that is out-right wrong (so far)...

```
~/trealla (devel) $ tpl
Trealla Prolog (c) Infradig 2020-2023, v2.27.20-1-gc049-dirty
?- use_module(library(clpb)).
   true.
?- sat(X*Y + X*Z), labeling([X,Y,Z]).
   X = 1, Y = X, Z = 0
;  X = 1, Y = X, Z = X
;  false.
?- 
```

The first solution is missing, as per below...

```
~/trealla (devel) $ scryer-prolog
?- use_module(library(clpb)).
   true.
?- sat(X*Y + X*Z), labeling([X,Y,Z]).
   X = 1, Y = 0, Z = 1
;  X = 1, Y = 1, Z = 0
;  X = 1, Y = 1, Z = 1.
?- 
```

**Comment by triska** [comment]
The choicepoint is also unexpected!

**Comment by infradig** [comment]
Gone from missing a solution to having one extra...

```
~/trealla (devel) $ tpl
Trealla Prolog (c) Infradig 2020-2023, v2.27.54-2-g9792
?- use_module(library(clpb)).
   true.
?-  sat(X*Y + X*Z), labeling([X,Y,Z]).
   X = 1, Y = 0, Z = Y
;  X = 1, Y = 0, Z = X
;  X = 1, Y = X, Z = 0
;  X = 1, Y = X, Z = X.
?- 
```

**Comment by triska** [comment]
I highly recommend first getting `clpb` right, and only then start looking at the `clpz` issue!

**Comment by infradig** [comment]
This `clpz` query is about as simple as it gets:

```
  ?- A#=B+C.
  false.
```

no hook involved, so it seems a good place to work on. Also, all remaining `clpz` errors seem to end up at the same place so i'm optimistic it's possibly the last stumbling block.

**Comment by triska** [comment]
I highly recommend you focus on `library(clpb)` first.

For the present issue, we can see a problem already without `labeling/1`:

<pre>
<b>?- sat(X*Y + X*Z).</b>
   <b>clpb:sat(X=\=X*Z#Z)</b>, unexpected.
</pre>

For comparison, Scryer Prolog yields the expected answer:

<pre>
<b>?- sat(X*Y + X*Z).</b>
   <b>X = 1</b>, clpb:sat(Y=\=Y*Z#Z).
</pre>

And a simpler goal, with Trealla:

<pre>
<b>?- clpb:parse_sat(X*Y + X*Z, Sat).</b>
   <b>Sat = X+Z</b>, clpb:sat(X=:=X),clpb:sat(Z=:=Z), unexpected.
</pre>

Whereas with Scryer, we get the expected answer:

<pre>
<b>?- clpb:parse_sat(X*Y + X*Z, Sat).</b>
   <b>Sat = X*Y+X*Z</b>, clpb:sat(X=:=X), clpb:sat(Y=:=Y), clpb:sat(Z=:=Z).
</pre>

**Comment by triska** [comment]
Systematically reducing it further, we have:

<pre>
<b>?- clpb:sat_rewrite(X*Y + X*Z, Sat). </b>
   Sat = X+Z, unexpected.
</pre>

The expected answer is:

<pre>
<b>?- clpb:sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X*Y+X*Z.
</pre>

**Comment by triska** [comment]
Here is a small sample program that shows the problem:

<pre>
sat_rewrite(V, V)       :- var(V), !.
sat_rewrite(P0*Q0, P*Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).
sat_rewrite(P0+Q0, P+Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).
</pre>

We get:

<pre>
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X+Z, unexpected.
</pre>

Whereas the expected result is:

<pre>
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X*Y+X*Z.
</pre>

So, something really elementary is currently broken.

**Comment by triska** [comment]
The issue persists:

<pre>
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X+Z, unexpected.
</pre>

**Comment by triska** [comment]
The issue still persists:

<pre>
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X+Z.
</pre>

**Comment by infradig** [comment]
I'm getting this...

~/trealla (devel) $ tpl
Trealla Prolog (c) Infradig 2020-2023, v2.29.2
?- use_module(library(clpb)).
   true.
?- clpb:parse_sat(X*Y + X*Z, Sat).
   Sat = X*Y+X*Z, clpb:sat(X=:=X),clpb:sat(Y=:=Y),clpb:sat(Z=:=Z).
?-
~/trealla (devel) $

Also, try adding -O0 to the command line as it seems to be a problem with
TCO's and attributes.

**Comment by triska** [comment]
Here is the full interaction, using the code I posted in https://github.com/trealla-prolog/trealla/issues/338#issuecomment-1747544829:

<pre>
<b>$ ./tpl 
Trealla Prolog (c) Infradig 2020-2023, v2.29.2
?- [user].</b>
sat_rewrite(V, V)       :- var(V), !.
sat_rewrite(P0*Q0, P*Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).
sat_rewrite(P0+Q0, P+Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).
<b>% Ctrl+d</b>
   true.
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X+Z.
</pre>

**No attributes occur in this example!** This is really basic code, it would be ridiculous to try to tackle `library(clpz)` as long as such elementary errors are present in the engine.

I can confirm that the problem is not present when I invoke Trealla with `-O0`.

**Comment by infradig** [comment]
Well there's something that interacts badly with TCO because -O0 just
disables that.

**Comment by infradig** [comment]
Fixed the TCO problem. It was overwriting the bb_b_put/2 backtracking info
in the environment.

**Comment by triska** [comment]
As far as I can tell, the issue persists:

<pre>
<b>$ ./tpl</b>
Trealla Prolog (c) Infradig 2020-2023, v2.29.5
<b>?- [user].</b>
sat_rewrite(V, V)       :- var(V), !.
sat_rewrite(P0*Q0, P*Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).
sat_rewrite(P0+Q0, P+Q) :- sat_rewrite(P0, P), sat_rewrite(Q0, Q).

   true.
<b>?- sat_rewrite(X*Y + X*Z, Sat).</b>
   Sat = X+Z.
</pre>

**Comment by infradig** [comment]
Correct I merely disabled TCO with clpb.pl while looking at other things.
That issue still needs addressing.

**Comment by infradig** [comment]
Ok, I think all the issues above are resolved excpet the original problem, which is back to missing one solution...

~/trealla (devel) $ tpl
Trealla Prolog (c) Infradig 2020-2023, v2.29.7
?- use_module(library(clpb)).
   true.
?- sat(X*Y + X*Z), labeling([X,Y,Z]).
   X = 1, Y = X, Z = 0
;  X = 1, Y = X, Z = X.
?-

**Comment by infradig** [comment]
Can you please recheck this one as well?

**Comment by triska** [comment]
I get:

<pre>
<b>?- sat(X*Y + X*Z), labeling([X,Y,Z]).</b>
   X = 1, Y = 1, Z = 0
;  X = 1, Y = 1, Z = 1
;  false.
</pre>

You can easily see that this is not correct, from the comparison with Scryer Prolog:

<pre>
<b>?- sat(X*Y + X*Z), labeling([X,Y,Z]).</b>
   X = 1, Y = 0, Z = 1
;  X = 1, Y = 1, Z = 0
;  X = 1, Y = 1, Z = 1.
</pre>

You can also use Trealla itself to see that either the first or the following is wrong, **because adding constraints succeeds where a generalization fails**:

<pre>
<b>?- X = 1, Y = 0, Z = 1, sat(X*Y + X*Z), labeling([X,Y,Z]).</b>
   X = 1, Y = 0, Z = 1.
</pre>

In short, we have:

<pre>
<b>?- sat(X*Y + X*Z), labeling([X,Y,Z]), Y = 0.</b>
   false.
</pre>

Whereas:

<pre>
<b>?- Y = 0, sat(X*Y + X*Z), labeling([X,Y,Z]).</b>
   Y = 0, X = 1, Z = 1
;  false.
</pre>

**Comment by jgarte** [comment]
> I highly recommend first getting `clpb` right, and only then start looking at the `clpz` issue!

Hi @triska Why is clpb a prerequisite for clpz? Just trying to understand for myself.

**Comment by triska** [comment]
> Why is clpb a prerequisite for clpz? Just trying to understand for myself.

In addition to the significantly larger amount of code (`clpz` is roughly 4 times as large as `clpb`), the `clpz` implementation uses features that `clpb` does not, notably **term expansion** to generate the code of the library in the first place.

We cannot point at a predicate that is generated when the library is loaded and say "this is about as simple as it gets", because that already presumes that term&nbsp;expansion and all related mechanisms work correctly. A working `clpb` helps ensure that more of the underlying system works correctly, and we can use it to reliably generate additional predicates.

**Comment by infradig** [comment]
However `clpb` uses the blackboard which is more of a complication, especially the backtraking part. While my testing of the blackboard shows it *seems* ok, I can't guarantee it's not the source of the problem.

**Comment by jgarte** [comment]
Hi @infradig what is the blackboard that you refer to? Not sure I am familiar with it.

**Comment by infradig** [comment]
The bb_XXX predicates.

**Comment by jgarte** [comment]
Got it

**Comment by infradig** [comment]
Fixed:
```
➜  trealla git:(main) ✗ tpl -g halt -f tests/issues/test338.pl
[1,0,1]
[1,1,0]
[1,1,1]
➜  trealla git:(main) ✗ scryer-prolog -g halt -f tests/issues/test338.pl
[1,0,1]
[1,1,0]
[1,1,1]
➜  trealla git:(main) ✗
```
